tools/deploy_firmware.py

```diff
@@ -0,0 +1,287 @@
+
+        self.rootdir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..")
+        self.artifacts_dir = os.path.join(self.rootdir, "build/{}".format(self.device))
+        """The directory where artifacts are stored. It's expected to contain
+        prplwrt-version, and the image file."""
+
+    def set_prompt(self, pexp: pexpect):
+        """Set the prompt to `serial_prompt` on a pexpect object."""
+        pexp.sendline("export PS1='{}'".format(self.serial_prompt))
+
+    def read_remote_version(self) -> List[str]:
+        """Read prplwrt-version on a remote device
+
+        Returns
+        -------
+        List[str]
+            The content of the prplwrt-version file on the device as a list of strings.
+        """
+        version = None
+        with pexpect.pxssh.pxssh(logfile=sys.stdout.buffer) as shell:
+            shell.login(self.name, self.username)
+            shell.sendline("cat /etc/prplwrt-version")
+            shell.expect("cat /etc/prplwrt-version")
+            shell.prompt()
+            version = shell.before.decode().splitlines()
+        return version
+
+    def read_artifacts_dir_version(self) -> List[str]:
+        """Reads the local prplwrt-version.
+
+        The version file is read from the artifacts (see `artifacts_dir`).
+
+        Returns
+        -------
+        List[str]
+            The content of the local prplwrt-version as a list of strings.
+        """
+        with open(os.path.join(self.artifacts_dir, "prplwrt-version")) as version_file:
+            return version_file.read().splitlines()
+
+    def needs_upgrade(self):
+        """Check if a device needs to be updated
+
+        The check is done by comparing prplwrt-version on the target
+        with a local one (see `read_artifacts_dir_version).
+        """
+        artifacts_dir_version = self.read_artifacts_dir_version()
+        remote_version = self.read_remote_version()
+        diff = difflib.unified_diff(artifacts_dir_version, remote_version,
+                                    fromfile='file1', tofile='file2')
+        diff_str = '\n'.join(diff)
+        if diff_str:
+            print(diff_str)
+        return bool(diff_str)
+
+    def reach(self, attempts: int = 5, wait: int = 5):
+        """Check if the device is reachable via SSH (and optionally try multiple times).
+
+        Parameters
+        ----------
+        attempts: int, optional
+            How many times we try before concluding that the device is unreachable.
+        wait: int, optional
+            The number of seconds to wait between attempts.
+
+        Returns
+        -------
+        bool
+            True if the device is reachable via SSH, false otherwise.
+        """
+        for _ in range(attempts):
+            try:
+                with pexpect.pxssh.pxssh() as shell:
+                    shell.login(self.name, self.username, login_timeout=5)
+                    return True
+            except pexpect.pxssh.ExceptionPxssh:
```

Comment:
I'm not sure retrying would always be effective. If the exception is (for example) "permission denied" or "password refused" there is no point in retrying and spending waiting time.
The exception can be parsed and only retry when we have an error that can be solved by retrying.

The source details all the possible exceptions on login:
https://pexpect.readthedocs.io/en/stable/_modules/pexpect/pxssh.html

Reply:
True, that would be an improvement (for later, it won't have any big impact in CI).

pxssh raises the same exception for all the cases though, so it won't be possible to have different except closes for the different cases. Is it what you were suggesting?
It would also be possible to distinguish the cases based on the exception message, but it's ugly and prone to fail between different versions of pxssh.